File: core/Core/Services/CardService.cs
using AutoMapper;
using Core.Exceptions;
using Core.Interfaces;
using Core.Models;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;

namespace Core.Services;

public class CardService(ICardRepository repository, IMapper mapper, ICacheService cacheService, IStatsService statsService) : Core.CardService.CardServiceBase
{
    public override async Task<CardModel> Add(AddCardRequest request, ServerCallContext context)
    {
        var card = mapper.Map<Card>(request);
        await repository.AddAsync(card);
        cacheService.Remove("BankStats");
        return mapper.Map<CardModel>(card);
    }

    public override async Task<GetAllCardsResponse> GetAll(GetAllRequest request, ServerCallContext context)
    {
        var cards = await repository.GetAllAsync(request.PageN, request.PageSize, request.OrderBy, request.IsDesc ?? false);
        return new GetAllCardsResponse
        {
            Cards = { mapper.Map<IEnumerable<CardModel>>(cards) }
        };
    }

    public override async Task<Empty> Delete(DeleteCardRequest request, ServerCallContext context)
    {
        var card = await repository.GetByIdAsync(request.CardId);
        if (card == null)
        {
            throw new NotFoundException("Card not found");
        }
        await repository.DeleteAsync(card.CardId);
        cacheService.Remove("BankStats");
        return new Empty();
    }

    public override async Task<Empty> Update(UpdateCardRequest request, ServerCallContext context)
    {
        var card = await repository.GetByIdAsync(request.CardId);
        if (card == null)
        {
            throw new NotFoundException("Card not found");
        }

        mapper.Map(request, card);

        await repository.UpdateAsync(card);
        cacheService.Remove("BankStats");

        return new Empty();
    }

    public override async Task<CardModel> GetById(GetCardByIdRequest request, ServerCallContext context)
    {
        var card = await repository.GetByIdAsync(request.CardId);
        if (card == null)
        {
            throw new NotFoundException("Card not found");
        }

        return mapper.Map<CardModel>(card);
    }

    public override async Task<Empty> DeleteBulk(DeleteCardBulkRequest request, ServerCallContext context)
    {
        var ids = request.Cards.Select(c => c.CardId).ToList();
        if (ids.Count == 0)
        {
            throw new ValidationException("No cards to delete");
        }
        var cards = await repository.GetByIdsAsync(ids);
        var foundCards = cards.Where(c => c != null).ToList();
        if (foundCards.Count != ids.Count)
        {
            throw new ValidationException("Some cards not found");
        }
        await repository.DeleteRangeAsync(foundCards!);
        cacheService.Remove("BankStats");
        return new Empty();
    }

    public override async Task<Empty> UpdateBulk(UpdateCardBulkRequest request, ServerCallContext context)
    {
        var cards = request.Cards.Select(mapper.Map<Card>).ToList();
        if (!cards.Any())
        {
            throw new ValidationException("No cards to update");
        }
        await repository.UpdateRangeAsync(cards);
        cacheService.Remove("BankStats");
        return new Empty();
    }

    public override async Task<Empty> AddBulk(AddCardBulkRequest request, ServerCallContext context)
    {
        var cards = request.Cards.Select(mapper.Map<Card>).ToList();
        await repository.AddRangeAsync(cards);
        cacheService.Remove("BankStats");
        return new Empty();
    }

    public override async Task<GetAllCardsResponse> GetByAccount(GetCardsByAccountRequest request, ServerCallContext context)
    {
        var cards = await repository.FindAsync(c => c.AccountId == request.AccountId);
        return new GetAllCardsResponse()
        {
            Cards = { mapper.Map<IEnumerable<CardModel>>(cards) }
        };
    }

    public override async Task<GetAllCardsResponse> GetByIds(GetCardByIdsRequest request, ServerCallContext context)
    {
        var cards = await repository.GetByIdsAsync(request.CardIds);
        return new GetAllCardsResponse()
        {
            Cards = { mapper.Map<IEnumerable<CardModel>>(cards) }
        };
    }

    public override async Task<CountResponse> GetCount(Empty request, ServerCallContext context)
    {
        var count = await repository.GetCountAsync();
        return new CountResponse()
        {
            Count = count
        };
    }

    public override async Task<CountResponse> GetCountByStatus(GetCardCountByStatus request, ServerCallContext context)
    {
        if (request.Status == "active")
        {
            var stats = await statsService.GetStatsAsync();
            return new CountResponse()
            {
                Count = stats.ActiveCardCount
            };
        }
        var count = await repository.GetCountAsync(c => c.Status == request.Status);
        return new CountResponse()
        {
            Count = count
        };
    }
}

File: orch-go/internal/services/card_service.go
package services

import (
	"context"
	"orch-go/internal/domain/card"
	"orch-go/internal/infrastructure/repository/card_repo"
)

type CardService struct {
	repo card_repo.Repository
}

func NewCardService(repo card_repo.Repository) *CardService {
	return &CardService{repo: repo}
}

func (s *CardService) GetCardById(ctx context.Context, id int32) (*card.Card, error) {
	return s.repo.GetById(ctx, id)
}

func (s *CardService) GetCardsByAccountId(ctx context.Context, accountId int32) ([]*card.Card, error) {
	return s.repo.GetByAccountId(ctx, accountId)
}

func (s *CardService) GetAllCards(ctx context.Context) ([]*card.Card, error) {
	return s.repo.GetAll(ctx)
}

func (s *CardService) CreateCard(ctx context.Context, card *card.Card) (*card.Card, error) {
	return s.repo.Add(ctx, card)
}

func (s *CardService) UpdateCard(ctx context.Context, card *card.Card) error {
	return s.repo.Update(ctx, card)
}

func (s *CardService) DeleteCard(ctx context.Context, id int32) error {
	return s.repo.Delete(ctx, id)
}

func (s *CardService) CreateCardBulk(ctx context.Context, cards []*card.Card) error {
	return s.repo.AddBulk(ctx, cards)
}

func (s *CardService) UpdateCardBulk(ctx context.Context, cards []*card.Card) error {
	return s.repo.UpdateBulk(ctx, cards)
}

func (s *CardService) DeleteCardBulk(ctx context.Context, ids []int32) error {
	return s.repo.DeleteBulk(ctx, ids)
}
